File: wagnerChess/player.py
from piece import Piece
from king import King
from queen import Queen
from bishop import Bishop
from knight import Knight
from rook import Rook
from pawn import Pawn

class Player(object):

    def __init__(self,color,boardRoot):
        
        if color not in ["white","black"]:
            raise Exception("invalid color")
        self.color=color

        self.root=boardRoot

        board=self.root["board"].getBoard()#board[","]

        if self.color=='white':

            self.root['pieces'].append(King(location=board["H,1"],root=self.root,color=self.color))
            self.root['pieces'].append(Queen(location=board["G,0"],root=self.root,color=self.color))
            self.root['pieces'].append(Bishop(board["H,2"],root=self.root,color=self.color))
            self.root['pieces'].append(Bishop(board["F,0"],root=self.root,color=self.color))
            self.root['pieces'].append(Knight(board["H,3"],root=self.root,color=self.color))
            self.root['pieces'].append(Knight(board["E,0"],root=self.root,color=self.color))
            self.root['pieces'].append(Rook(board["H,4"],root=self.root,color=self.color))
            self.root['pieces'].append(Rook(board["D,0"],root=self.root,color=self.color))
            
            self.root['pieces'].append(Pawn(board["H,0"],root=self.root,color=self.color))
            self.root['pieces'].append(Pawn(board["F,1"],root=self.root,color=self.color))
            self.root['pieces'].append(Pawn(board["G,2"],root=self.root,color=self.color))
            self.root['pieces'].append(Pawn(board["E,3"],root=self.root,color=self.color))
            self.root['pieces'].append(Pawn(board["F,2"],root=self.root,color=self.color))
            self.root['pieces'].append(Pawn(board["G,3"],root=self.root,color=self.color))
            self.root['pieces'].append(Pawn(board["G,1"],root=self.root,color=self.color))
            self.root['pieces'].append(Pawn(board["E,1"],root=self.root,color=self.color))
            
        else:
            self.root['pieces'].append(King(location=board["B,7"],root=self.root,color=self.color))
            self.root['pieces'].append(Queen(location=board["A,6"],root=self.root,color=self.color))
            self.root['pieces'].append(Bishop(board["A,5"],root=self.root,color=self.color))
            self.root['pieces'].append(Bishop(board["C,7"],root=self.root,color=self.color))
            self.root['pieces'].append(Knight(board["A,4"],root=self.root,color=self.color))
            self.root['pieces'].append(Knight(board["D,7"],root=self.root,color=self.color))
            self.root['pieces'].append(Rook(board["A,3"],root=self.root,color=self.color))
            self.root['pieces'].append(Rook(board["E,7"],root=self.root,color=self.color))
                                 
            self.root['pieces'].append(Pawn(board["A,7"],root=self.root,color=self.color))
            self.root['pieces'].append(Pawn(board["C,6"],root=self.root,color=self.color))
            self.root['pieces'].append(Pawn(board["B,5"],root=self.root,color=self.color))
            self.root['pieces'].append(Pawn(board["B,4"],root=self.root,color=self.color))
            self.root['pieces'].append(Pawn(board["C,5"],root=self.root,color=self.color))
            self.root['pieces'].append(Pawn(board["D,4"],root=self.root,color=self.color))
            self.root['pieces'].append(Pawn(board["B,6"],root=self.root,color=self.color))
            self.root['pieces'].append(Pawn(board["D,6"],root=self.root,color=self.color))
            

    #True/False, [] of pieces attacking opposing King
    def hasCheck(self):
        #collect all self.color pieces from the Collection
        checkingPieces=[]
        pieces=[]
        opposingKing=None

        for p in self.root['pieces']:
            print ("test:"+p.toString())
        
        for p in self.root['pieces']:
            if p.getColor()==self.color:
                pieces.append(p)
                print ("my piece:"+p.toString())
                
            elif p.getColor()!=self.color and p.getPieceType()=="king":
                opposingKing=p
                print ("opposing king: "+p.toString())


        print ("0")
        
        #see if each piece can attack the other King
        for p in pieces:
            print ("Can this threaten the king?"+p.toString()+"@"+p.getLocation().locString())
            threatenedPieces=p.canThreatenList()
            
            print("ThreatenedPieces=")
            for tp in threatenedPieces:
                print(tp.getLocation().toString()+'@'+tp.getLocation().locString())

            
            if opposingKing in threatenedPieces:
                print ("5")
                checkingPieces.append(p)

        print ("6")
        if len(checkingPieces)>0:
            return True,checkingPieces
        else:
            return False,checkingPieces

File: cWagnerChess/player.c
#include "player.h"

PlayerPtr initPlayer(char* color){
  PlayerPtr p = (PlayerPtr)malloc(sizeof(Player));
  p->color=color;
  p->toString=toStringPlayer;
  p->getColor=playerGetColor;

  p->pieces[KING] = initPiece(KING,color);
  p->pieces[QUEEN] = initPiece(QUEEN,color);
  p->pieces[BISHOP] = initPiece(BISHOP,color);
  p->pieces[KNIGHT] = initPiece(KNIGHT,color);
  p->pieces[ROOK] = initPiece(ROOK,color);
  p->pieces[PAWN] = initPiece(PAWN,color);
  p->pieces[PAWN+1] = initPiece(PAWN,color);
  p->pieces[PAWN+2] = initPiece(PAWN,color);
  p->pieces[PAWN+3] = initPiece(PAWN,color);
  p->pieces[PAWN+4] = initPiece(PAWN,color);
  p->pieces[PAWN+5] = initPiece(PAWN,color);
  p->pieces[PAWN+6] = initPiece(PAWN,color);
  p->pieces[PAWN+7] = initPiece(PAWN,color);
  //
  p->pieces[PAWN+8] = initPiece(BISHOP,color);
  p->pieces[PAWN+9] = initPiece(KNIGHT,color);
  p->pieces[PAWN+10] = initPiece(ROOK,color);
  return p;

}

char* toStringPlayer(PlayerPtr p){

  char *t="",*s=(char*)malloc(50);
  int i=0;
  for(i=0; i<= PAWN+10 ; i++){
    sprintf(s,"%s\n",p->pieces[i]->toString(p->pieces[i]));
    t=mystrcat(t,s);
  }

  return t;
}

char* playerGetColor(PlayerPtr p){
  return p->color;
}
